compiler/p/codegen/OMRTreeEvaluator.cpp

```diff
@@ -5189,6 +5191,232 @@ TR::Register *OMR::Power::TreeEvaluator::directCallEvaluator(TR::Node *node, TR:
    }
 
+static TR::Register *inlineSimpleAtomicUpdate(TR::Node *node, bool isAddOp, bool isLong, bool isGetThenUpdate, TR::CodeGenerator *cg)
+   {
+   TR::Node *valueAddrChild = node->getFirstChild();
+   TR::Node *deltaChild = NULL;
+   TR::Register *valueAddrReg = cg->evaluate(valueAddrChild);
+   TR::Register *deltaReg = NULL;
+   TR::Register *resultReg = cg->allocateRegister();
+   TR::Register *cndReg = cg->allocateRegister(TR_CCR);
+   TR::Register *tempReg = NULL;
+   bool isDeltaConstant = false;
+   bool isDeltaImmediate = false;
+   bool isDeltaImmediateShifted = false;
+
+   int32_t numDeps = 4;
+
+   int32_t delta = 0;
+
+   // Memory barrier --- NOTE: we should be able to do a test upfront to save this barrier,
+   //                          but Hursley advised to be conservative due to lack of specification.
+   generateInstruction(cg, TR::InstOpCode::lwsync, node);
+
+   TR::LabelSymbol *doneLabel = TR::LabelSymbol::create(cg->trHeapMemory(),cg);
+   TR::LabelSymbol *loopLabel = TR::LabelSymbol::create(cg->trHeapMemory(),cg);
+
+   loopLabel->setStartInternalControlFlow();
+   deltaChild = node->getSecondChild();
+
+   if (deltaChild->getOpCode().isLoadConst()
+          && !deltaChild->getRegister()
+          && deltaChild->getDataType() == TR::Int32)
+      {
+      delta = (int32_t)(deltaChild->getInt());
+      isDeltaConstant = true;
+
+      //determine if the constant can be represented as an immediate
+      if (delta <= UPPER_IMMED && delta >= LOWER_IMMED)
+         {
+         // avoid evaluating immediates for add operations
+         isDeltaImmediate = true;
+         }
+      else if (delta & 0xFFFF == 0 && (delta & 0xFFFF0000) >> 16 <= UPPER_IMMED && (delta & 0xFFFF0000) >> 16 >= LOWER_IMMED)
+         {
+         // avoid evaluating shifted immediates for add operations
+         isDeltaImmediate = true;
+         isDeltaImmediateShifted = true;
+         }
+      else
+         {
+         // evaluate non-immediate constants since there may be reuse
+         // and they have to go into a reg anyway
+         tempReg = cg->evaluate(deltaChild);
+         }
+      }
+   else
+      {
+      tempReg = cg->evaluate(deltaChild);
+      }
+
+   generateLabelInstruction(cg, TR::InstOpCode::label, node, loopLabel);
+
+   deltaReg = cg->allocateRegister();
+   if (isDeltaImmediate && !isAddOp)
+      {
+      // If argument is immediate, but the operation is not an add,
+      // the value must still be loaded into a register
+      // If argument is an immediate value and operation is an add,
+      // it will be used as an immediate operand in an add immediate instruction
+      loadConstant(cg, node, delta, deltaReg);
+      }
+   else if (!isDeltaImmediate)
+      {
+      // For non-constant arguments, use evaluated register
+      // For non-immediate constants, evaluate since they may be re-used
+      numDeps++;
+      generateTrg1Src1Instruction(cg, TR::InstOpCode::mr, node, deltaReg, tempReg);
```

Comment:
Should explicitly test if it will be re-used ... and you potentially can save this copy instruction.

Reply:
@zl-wang Julian, I assume you mean only for the non-immediate constant case.  I think you're suggesting that in the case of a non-immediate constant, it should evaluate the constant in the loop, but for a non-constant, it should be evaluated outside of the loop, the value copied into another register to using an `mr`.  Do I have that right?

Now that I'm thinking about it, why does this code try to reuse the register that holds delta to hold the sum. . . ?

Reply:
@hzongaro  certainly the new sequence is better.  previously, my comment, looking at the mr instruction, meant to say why not testing the delta child referenceCount to decide if delta can be reused.  In your new sequence, if  deltaChild->getReferenceCount() == 1, you can even eliminate the creation of GPR_88 and use GPR_99 in its place.  But this is not strictly necessary.

Reply:
Revised generated code so that the `mr` in the loop is no longer used.  Changed in commit https://github.com/eclipse/omr/pull/3018/commits/e8493aa681a66b85e9d83cca9cbe1f7dfdda4c1b